gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/ImmunizationMapper.java

```diff
@@ -0,0 +1,171 @@
+
+            return createImmunization(id, identifier, note, vaccineExtension, date, recordedTimeExtension, patient, encounter);
+        }
+
+        return null;
+    }
+
+    private Extension getRecordedTimeExtension(RCMRMT030101UK04EhrComposition ehrComposition) {
+        if (ehrComposition.getAuthor() != null) {
+            return new Extension()
+                .setValue(new StringType(ehrComposition.getAuthor().getTime().getValue()));
+        } else if (ehrComposition.getEffectiveTime() != null) {
+            return new Extension()
+                .setValue(new StringType(DateFormatUtil.parse(ehrComposition.getAvailabilityTime().getValue()).asStringValue()));
+        }
+
+        return null;
+    }
+
+    private Identifier getIdentifier(String id) {
+        return new Identifier()
+            .setSystem(IDENTIFIER_SYSTEM) // TODO: concatenate source practice org id to URL (NIAD-2021)
+            .setValue(id);
+    }
+
+    private boolean hasImmunizationCode(RCMRMT030101UK04ObservationStatement observationStatement) {
+        String snomedCode = observationStatement.getCode().getCodeSystem();
+
+        return IMMUNIZATION_SNOMED_CODE.equals(snomedCode);
+    }
+
+    private Extension createVaccineProcedureExtension(RCMRMT030101UK04ObservationStatement observationStatement) {
+        return new Extension()
+            .setUrl(VACCINE_PROCEDURE_URL)
+            .setValue(new Extension()
+                .setUrl(VACCINATION_CODING_EXTENSION_URL)
+                .setValue(codeableConceptMapper.mapToCodeableConcept(observationStatement.getCode()))
+            );
+    }
+
+    private Date getObservationDate(RCMRMT030101UK04ObservationStatement observationStatement) {
+        if (observationStatement.getEffectiveTime().getLow() != null && observationStatement.getEffectiveTime().getHigh() != null) {
+            return DateFormatUtil.parse(observationStatement.getEffectiveTime().getLow().getValue()).getValue();
+        } else if (observationStatement.getEffectiveTime().getCenter() != null) {
+            return DateFormatUtil.parse(observationStatement.getEffectiveTime().getCenter().getValue()).getValue();
+        } else if (observationStatement.getEffectiveTime().getLow() != null) {
+            return DateFormatUtil.parse(observationStatement.getEffectiveTime().getLow().getValue()).getValue();
+        } else if (observationStatement.getEffectiveTime().getHigh() != null
+            && observationStatement.getAvailabilityTime() != null
+            && observationStatement.getAvailabilityTime().getValue() != null) {
+            return DateFormatUtil.parse(observationStatement.getAvailabilityTime().getValue()).getValue();
+        } else if (observationStatement.getEffectiveTime() == null && observationStatement.getAvailabilityTime().getValue() != null) {
+            return DateFormatUtil.parse(observationStatement.getAvailabilityTime().getValue()).getValue();
```

Comment:
Do me a favour and check each of these values with nullFlavor="UNK" insetad of a value. I suspect this might throw an error. My current PR has some functionality which checks for this if you need a reference https://github.com/NHSDigital/nia-patient-switching-standard-adaptor/pull/57/files#diff-fadb5d3a335b0ecd8a6396f0e489d8ff26a4ad6d8f517e561c710d17f9da8600R47